modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFunction.java

```diff
@@ -122,9 +149,17 @@ void analyze(ScriptRoot scriptRoot) {
         block.analyze(scriptRoot, functionScope.newLocalScope());
         methodEscape = block.methodEscape;
 
-        if (!methodEscape && returnType != void.class) {
-            throw createError(new IllegalArgumentException("Not all paths provide a return value for method [" + name + "]."));
+        if (methodEscape == false && isAutoReturnEnabled == false && returnType != void.class) {
+            throw createError(new IllegalArgumentException("not all paths provide a return value " +
+                    "for function [" + name + "] with [" + typeParameters.size() + "] parameters"));
+        }
+
+        // TODO: do not specialize for execute
```

Comment:
no need to change, but I usually do something like 
`// TODO: do not specialize for execute, see #51841`

Reply:
Got it. Will do for the next instance.